File: 2020/csharp/Day02.cs
using System.Text;
using System.Text.RegularExpressions;

namespace AdventOfCode.CSharp.Y2020;

public static class Day02
{
    public static void Run(byte[] input, Action<int, string> output)
    {
        var Input = Encoding.UTF8
            .GetString(input)
            .Split('\n', StringSplitOptions.RemoveEmptyEntries);

        int Part1()
        {
            bool checkValid(string line)
            {
                var m = Regex.Match(line, @"([0-9]+)\-([0-9]+) ([a-z])\: ([a-z]+)");
                if (!m.Success)
                {
                    throw new InvalidOperationException($"Line doesn't match: {line}");
                }

                var min = int.Parse(m.Groups[1].Value);
                var max = int.Parse(m.Groups[2].Value);
                var c = m.Groups[3].Value[0];
                var p = m.Groups[4].Value;

                var ccount = p.Where(x => x == c).Count();

                return ccount >= min && ccount <= max;
            }

            var numbers =
                from line in Input
                where checkValid(line)
                select line;

            return numbers.Count();
        }

        int Part2()
        {
            bool checkValid(string line)
            {
                var m = Regex.Match(line, @"([0-9]+)\-([0-9]+) ([a-z])\: ([a-z]+)");
                if (!m.Success)
                {
                    throw new InvalidOperationException($"Line doesn't match: {line}");
                }

                var p1 = int.Parse(m.Groups[1].Value) - 1;
                var p2 = int.Parse(m.Groups[2].Value) - 1;
                var c = m.Groups[3].Value[0];
                var password = m.Groups[4].Value;

                return password[p1] == c ^ password[p2] == c;
            }

            var numbers =
                from line in Input
                where checkValid(line)
                select line;

            return numbers.Count();
        }

        output(1, Part1().ToString());
        output(2, Part2().ToString());
    }
}

File: 2020/fsharp/day02.fs
namespace AdventOfCode.FSharp.Y2020

module Day02 =
    open AdventOfCode.FSharp.Util

    type Pline =
        { min: int
          max: int
          c: char
          input: string }

    let parse line =
        match line with
        | Regex @"(\d+)\-(\d+) ([a-z])\: ([a-z]+)" [ mint; maxt; ct; input ] ->
            Some
                { min = int mint
                  max = int maxt
                  c = ct.[0]
                  input = input }
        | _ -> None

    let part1 input =
        let valid p =
            let ccount =
                p.input |> Seq.where (fun x -> x = p.c) |> Seq.length

            p.min <= ccount && ccount <= p.max

        input
        |> splitLine
        |> Seq.choose parse
        |> Seq.where valid
        |> Seq.length
        |> string

    let part2 input =
        let valid p =
            (p.input.[p.min - 1] = p.c) <> (p.input.[p.max - 1] = p.c)

        input
        |> splitLine
        |> Seq.choose parse
        |> Seq.where valid
        |> Seq.length
        |> string
